File: prolog/state.pl
% state.pl - with constraint additions and fixes
:- module(state, [
    player_location/1, inventory/1, door_state/3,
    has_piece/2, object_moved/1, puzzle_solved/1,
    move_player/1, pick_key/1, pick_piece/1, move_object/1,
    solve_puzzle/1, unlock_door/2, has_key/1, remove_key/1,
    drop_key/1, drop_piece/1, key_dropped/2, piece_dropped/3
]).

:- dynamic player_location/1.
:- dynamic inventory/1.
:- dynamic door_state/3.
:- dynamic has_piece/2.
:- dynamic object_moved/1.
:- dynamic puzzle_solved/1.
:- dynamic key_dropped/2.    % key_dropped(Key, Room)
:- dynamic piece_dropped/3.  % piece_dropped(Piece, Puzzle, Room)
:- dynamic has_key/1.

:- use_module(facts).
:- use_module(constraints).
:- use_module(adversary).

% Para evitar warnings de predicados discontiguos
:- discontiguous pick_key/1, pick_piece/1.

% Initial state
player_location(a).
inventory([]).

% Initialize door states
:- forall(facts:door(X, Y, State), assertz(door_state(X, Y, State))).

% Move player between rooms with constraint checks

% Add to facts.pl - Add the final room predicate
:- dynamic final_room/1.

% Update module exports in facts.pl
:- module(facts, [
    room/1, door/3, key_in_room/2,
    object_in_room/2, hides_piece/3, piece/2,
    piece_location/3, puzzle_room/2,
    door_requirements/3, puzzle/1,
    piece_in_room/3, final_room/1,
    % Dynamic definition functions
    add_room/1, add_door/3, add_key/2,
    add_object/2, add_puzzle/1, add_piece/2,
    hide_piece/3, set_puzzle_room/2,
    set_door_requirements/3, set_final_room/1,
    % Game configuration functions
    clear_game_data/0,
    load_predefined_game/0,
    create_custom_game/0
]).

% Add the new function to set the final room
set_final_room(Room) :-
    retractall(final_room(_)),
    assertz(final_room(Room)).

% Update clear_game_data in facts.pl to also clear final_room
clear_game_data :-
    retractall(room(_)),
    retractall(door(_, _, _)),
    retractall(key_in_room(_, _)),
    retractall(object_in_room(_, _)),
    retractall(hides_piece(_, _, _)),
    retractall(piece(_, _)),
    retractall(piece_location(_, _, _)),
    retractall(puzzle_room(_, _)),
    retractall(door_requirements(_, _, _)),
    retractall(puzzle(_)),
    retractall(piece_in_room(_, _, _)),
    retractall(final_room(_)).

% Update load_predefined_game to set the final room
load_predefined_game :-
    clear_game_data,
    
    % Add rooms
    add_room(a), add_room(b), add_room(c), add_room(d),
    
    % Set final room
    set_final_room(d),
    
    % Rest of the predefined game setup...
    % [Keep all the existing code here]

% Update create_custom_game to include final room setup after door requirements
custom_game_door_requirements :-
    writeln('Finally, let\'s set the requirements for each door.'),
    writeln('Format: from_room-to_room-[req1,req2,...] or "done":'),
    writeln('Requirements can be has_key(KeyName) or puzzle_solved(PuzzleName)'),
    read(ReqInput),
    (ReqInput = done ->
        writeln('Door requirements setup complete.'),
        custom_game_final_room
    ;
        ReqInput = From-To-Reqs,
        set_door_requirements(From, To, Reqs),
        writeln('Door requirement set. Enter another door requirement or "done":'),
        custom_game_door_requirements
    ).

% Add new function to set final room in custom game
custom_game_final_room :-
    writeln('Finally, which room is the final goal (exit) room?'),
    findall(R, room(R), Rooms),
    format('Available rooms: ~w~n', [Rooms]),
    read(FinalRoom),
    (room(FinalRoom) ->
        set_final_room(FinalRoom),
        writeln('Final room set successfully!'),
        writeln('Custom game created successfully!')
    ;
        writeln('Invalid room. Please choose from the available rooms:'),
        format('~w~n', [Rooms]),
        custom_game_final_room
    ).

% Update state.pl - Modify move_player to check for win condition
% Add to beginning of state.pl module declarations
:- use_module(facts).

% Update move_player predicate
move_player(NewRoom) :-
    constraints:check_move_limit,
    player_location(Current),
    (door_state(Current, NewRoom, unlocked) ; door_state(NewRoom, Current, unlocked)),
    % Update counters FIRST before changing location
    constraints:increment_move_count,
    constraints:increment_room_turn(NewRoom),
    % Now update location
    retract(player_location(Current)),
    assertz(player_location(NewRoom)),
    % Show move confirmation
    format("You moved from ~w to ~w.~n", [Current, NewRoom]),
    (facts:game_mode(adversary) ->
        adversary:move_guard
    ;
        true
    ),
    
    % Check if this is the final room
    (facts:final_room(NewRoom) ->
        constraints:move_count(Moves),
        writeln("*********************************************"),
        writeln("*   CONGRATULATIONS! YOU HAVE ESCAPED!     *"),
        writeln("*********************************************"),
        format("You reached the exit room ~w in ~w moves!~n", [NewRoom, Moves]),
        writeln("Type 'init_game.' to play again.")
    ;
        % Not final room, continue normally
        true
    ),
    !.

% Check if player has a key
has_key(Key) :-
    inventory(Inv),
    member(Key, Inv).

% Remove a key from inventory
remove_key(Key) :-
    inventory(Inv),
    select(Key, Inv, NewInv),  % Remove key from inventory
    retract(inventory(Inv)),
    assertz(inventory(NewInv)),
    format("Used and removed key: ~w~n", [Key]).

% Pick up a key with constraint check
pick_key(Key) :-
    constraints:check_inventory_limit(key),  % Check inventory limit
    player_location(Room),
    (facts:key_in_room(Room, Key) ; key_dropped(Key, Room)),  % Verificar ambas fuentes
    inventory(Inv),
    \+ member(Key, Inv),
    % Si la llave había sido dejada, eliminar ese hecho
    (key_dropped(Key, Room) -> retract(key_dropped(Key, Room)) ; true),
    retract(inventory(Inv)),
    assertz(inventory([Key | Inv])),
    format("You have picked up key: ~w~n", [Key]),
    !.

pick_key(Key) :-
    player_location(Room),
    (facts:key_in_room(Room, Key) ; key_dropped(Key, Room)),
    inventory(Inv),
    member(Key, Inv),
    format("You already have key ~w.~n", [Key]),
    !.
    
pick_key(Key) :-
    format("Cannot pick up key ~w here.~n", [Key]).

% Pick up a puzzle piece with constraint check
pick_piece(Piece) :-
    constraints:check_inventory_limit(piece),  % Check inventory limit
    player_location(Room),
    (facts:piece_in_room(Room, Piece, Puzzle) ; piece_dropped(Piece, Puzzle, Room)),
    \+ has_piece(Puzzle, Piece),
    % Si la pieza había sido dejada, eliminar ese hecho
    (piece_dropped(Piece, Puzzle, Room) -> retract(piece_dropped(Piece, Puzzle, Room)) ; true),
    assertz(has_piece(Puzzle, Piece)),
    format("You have picked up piece ~w of puzzle ~w!~n", [Piece, Puzzle]),
    !.

pick_piece(Piece) :-
    player_location(Room),
    (facts:piece_in_room(Room, Piece, Puzzle) ; piece_dropped(Piece, Puzzle, Room)),
    has_piece(Puzzle, Piece),
    format("You already have piece ~w of puzzle ~w.~n", [Piece, Puzzle]),
    !.
    
pick_piece(Piece) :-
    format("Cannot pick up piece ~w here.~n", [Piece]).

% Dejar una llave en la habitación actual
drop_key(Key) :-
    player_location(Room),
    has_key(Key),
    inventory(Inv),
    select(Key, Inv, NewInv),    % Eliminar la llave del inventario
    retract(inventory(Inv)),
    assertz(inventory(NewInv)),
    assertz(key_dropped(Key, Room)),
    format("Has dejado la llave ~w en la habitación ~w.~n", [Key, Room]),
    !.
drop_key(Key) :-
    \+ has_key(Key),
    format("No tienes la llave ~w en tu inventario.~n", [Key]).

% Dejar una pieza de puzzle en la habitación actual
drop_piece(Piece) :-
    player_location(Room),
    % Buscar a qué puzzle pertenece la pieza
    has_piece(Puzzle, Piece),
    retract(has_piece(Puzzle, Piece)),
    assertz(piece_dropped(Piece, Puzzle, Room)),
    format("Has dejado la pieza ~w del puzzle ~w en la habitación ~w.~n", [Piece, Puzzle, Room]),
    !.
drop_piece(Piece) :-
    format("No tienes la pieza ~w en tu inventario.~n", [Piece]).

% Move an object to find hidden pieces
move_object(Object) :-
    player_location(Room),
    facts:object_in_room(Room, Object),
    \+ object_moved(Object),
    assertz(object_moved(Object)),
    % Check if moving reveals a puzzle piece
    ( facts:hides_piece(Object, Puzzle, Piece) ->
        ( constraints:check_inventory_limit(piece) ->
            assertz(has_piece(Puzzle, Piece)),
            format("You moved ~w and found piece ~w of puzzle ~w!~n", [Object, Piece, Puzzle])
        ;
            format("You moved ~w and found piece ~w of puzzle ~w, but your inventory is full!~n", [Object, Piece, Puzzle])
        )
    ;
        format("You moved ~w but found nothing interesting.~n", [Object])
    ),
    !.
move_object(Object) :-
    player_location(Room),
    facts:object_in_room(Room, Object),
    object_moved(Object),
    format("You've already examined ~w.~n", [Object]),
    !.
move_object(Object) :-
    format("There is no ~w here to move.~n", [Object]).

% Solve a puzzle - now removes pieces from inventory
solve_puzzle(Puzzle) :-
    player_location(Room),
    facts:puzzle_room(Puzzle, Room),
    \+ puzzle_solved(Puzzle),
    % Get all pieces needed for this puzzle
    findall(Piece, facts:piece(Puzzle, Piece), AllPieces),
    % Check if player has all necessary pieces
    forall(member(P, AllPieces), has_piece(Puzzle, P)),
    % Remove all pieces from inventory
    forall(member(P, AllPieces), retract(has_piece(Puzzle, P))),
    % Mark puzzle as solved
    assertz(puzzle_solved(Puzzle)),
    format("Congratulations! You have solved puzzle ~w! The pieces have been consumed.~n", [Puzzle]),
    !.
solve_puzzle(Puzzle) :-
    player_location(Room),
    facts:puzzle_room(Puzzle, Room),
    \+ puzzle_solved(Puzzle),
    format("You don't have all the pieces needed to solve puzzle ~w.~n", [Puzzle]),
    !.
solve_puzzle(Puzzle) :-
    player_location(Room),
    facts:puzzle(Puzzle),
    \+ facts:puzzle_room(Puzzle, Room),
    format("Puzzle ~w cannot be solved in this room.~n", [Puzzle]),
    !.
solve_puzzle(Puzzle) :-
    puzzle_solved(Puzzle),
    format("Puzzle ~w is already solved.~n", [Puzzle]),
    !.
solve_puzzle(Puzzle) :-
    format("Unknown puzzle: ~w~n", [Puzzle]).

% Dynamic door unlocking system - now consumes keys
unlock_door(From, To) :-
    door_state(From, To, locked),
    % Check if there are requirements for this door
    facts:door_requirements(From, To, Requirements),
    % Check if all requirements are met
    check_all_requirements(Requirements),
    % Consume keys used for unlocking
    consume_keys(Requirements),
    % Update door state
    retract(door_state(From, To, locked)),
    assertz(door_state(From, To, unlocked)),
    retract(door_state(To, From, locked)),
    assertz(door_state(To, From, unlocked)),
    format("Door between rooms ~w and ~w has been unlocked!~n", [From, To]),
    !.
unlock_door(From, To) :-
    door_state(From, To, unlocked),
    format("The door between rooms ~w and ~w is already unlocked.~n", [From, To]),
    !.
unlock_door(From, To) :-
    facts:door_requirements(From, To, Requirements),
    format("You cannot unlock this door yet. Requirements: ~w~n", [Requirements]),
    !.
unlock_door(From, To) :-
    format("There is no direct door between rooms ~w and ~w.~n", [From, To]).

% Consume keys used for unlocking
consume_keys([]).
consume_keys([has_key(Key)|Rest]) :-
    remove_key(Key),
    consume_keys(Rest).
consume_keys([_|Rest]) :-  % Skip non-key requirements
    consume_keys(Rest).

% Check if all requirements for a door are met
check_all_requirements([]).
check_all_requirements([Req|Rest]) :-
    check_requirement(Req),
    check_all_requirements(Rest).

% Check individual requirements
check_requirement(has_key(Key)) :-
    has_key(Key), !.
check_requirement(puzzle_solved(Puzzle)) :-
    puzzle_solved(Puzzle), !.
check_requirement(Req) :-
    format("Requirement not met: ~w~n", [Req]),
    fail.

File: python/integration/State.py
class StateManager:
    def __init__(self, prolog_bridge):
        self.prolog = prolog_bridge.prolog
        self.collected_items = {
            'keys': set(),
            'pieces': set()
        }

    def player_location(self):
        """Get current player location"""
        result = list(self.prolog.query("state:player_location(Room)"))
        return result[0]['Room'] if result else None
    
    def guard_location(self):
        """Get current guard location"""
        result = list(self.prolog.query("adversary:guard_location(GuardRoom)"))
        return result[0]['GuardRoom'] if result else None

    def inventory(self):
        """Get current inventory"""
        
        result = list(self.prolog.query("state:inventory(Inv)"))
        print("invez", result)
        return result[0]['Inv'] if result else []

    def door_state(self, room1, room2):
        """Get door state between two rooms"""
        result = list(self.prolog.query(f"state:door_state({room1}, {room2}, State)"))
        if result:
            return result[0]['State']
        # Check reverse direction if not found
        result = list(self.prolog.query(f"state:door_state({room2}, {room1}, State)"))
        return result[0]['State'] if result else None

    def has_piece(self, puzzle, piece):
        """Check if player has a puzzle piece"""
        return bool(list(self.prolog.query(f"state:has_piece({puzzle}, {piece})")))

    def object_moved(self, obj):
        """Check if an object has been moved"""
        return bool(list(self.prolog.query(f"state:object_moved({obj})")))

    def puzzle_solved(self, puzzle):
        """Check if a puzzle is solved"""
        return bool(list(self.prolog.query(f"state:puzzle_solved({puzzle})")))

    def move_player(self, new_room):
        """Move player to new room"""
        return bool(list(self.prolog.query(f"state:move_player({new_room})")))

    def pick_key(self, key):
        """Pick up a key"""
        query = f"""
            with_output_to(
                codes(Codes),
                state:pick_key({key})
            ),
            atom_codes(Output, Codes)
        """
        result = list(self.prolog.query(query))[0]['Output']
        result = result.strip()  # Elimina espacios y saltos de línea
        
        expected_message = f"You have picked up key: {key}"
        if result == expected_message:
            self.collected_items['keys'].add(key)
            print("temps agre", self.collected_items)
            return True
        return False


    def pick_piece(self, piece):
        """Pick up a puzzle piece"""
        query = f"""
                with_output_to(
                    codes(Codes),
                    state:pick_piece({piece})
                ),
                atom_codes(Output, Codes)
            """
        result = list(self.prolog.query(query))[0]['Output']
        result = result.strip()  # Elimina espacios y saltos de línea
        if "You have picked up" in result:
            self.collected_items['pieces'].add(piece)
            return True
        else:
            return False

    def move_object(self, obj):
        """Move an object in the room"""
        return bool(list(self.prolog.query(f"state:move_object({obj})")))

    def solve_puzzle(self, puzzle):
        """Solve a puzzle"""
        query = f"""
                with_output_to(
                    codes(Codes),
                    state:solve_puzzle({puzzle})
                ),
                atom_codes(Output, Codes)
            """
        result = list(self.prolog.query(query))[0]['Output']
        result = result.strip()  # Elimina espacios y saltos de línea    
        print("puzz", result)    
        if "Congratulations" in result or "already" in result:
            return True
        else:
            return False

    def unlock_door(self, from_room, to_room):
        """Unlock a door between rooms"""
        query = f"""
                with_output_to(
                    codes(Codes),
                    state:unlock_door({from_room}, {to_room})
                ),
                atom_codes(Output, Codes)
            """
        result = list(self.prolog.query(query))[0]['Output']
        result = result.strip()  # Elimina espacios y saltos de línea        
        # Verificamos si el resultado contiene "has been unlocked"
        print("resultaopuer",result)
        if "has been unlocked" in result:
            return True
        else:
            return False


    def has_key(self, key):
        """Check if player has a key"""
        return bool(list(self.prolog.query(f"state:has_key({key})")))

    def remove_key(self, key):
        """Remove a key from inventory"""
        return bool(list(self.prolog.query(f"state:remove_key({key})")))

    def drop_key(self, key):
        """Drop a key in current room"""
        query = f"""
                with_output_to(
                    codes(Codes),
                    state:drop_key({key})
                ),
                atom_codes(Output, Codes)
            """
        result = list(self.prolog.query(query))[0]['Output']
        result = result.strip()  # Elimina espacios y saltos de línea        
        if "No tienes" in result:
            return False
        else:
            self.collected_items['keys'].discard(key)
            return True


    def drop_piece(self, piece):
        """Drop a puzzle piece in current room"""
        query = f"""
                with_output_to(
                    codes(Codes),
                    state:drop_piece({piece})
                ),
                atom_codes(Output, Codes)
            """
        result = list(self.prolog.query(query))[0]['Output']
        result = result.strip()  # Elimina espacios y saltos de línea 
        if "No tienes" in result:
            return False
            
        else:
            self.collected_items['pieces'].discard(piece)
            return True

    def key_dropped(self, key, room):
        """Check if a key was dropped in a room"""
        return bool(list(self.prolog.query(f"state:key_dropped({key}, {room})")))

    def piece_dropped(self, piece, puzzle, room):
        """Check if a puzzle piece was dropped in a room"""
        return bool(list(self.prolog.query(f"state:piece_dropped({piece}, {puzzle}, {room})")))

    def get_current_state(self):
        """Get comprehensive game state"""
        result = list(self.prolog.query("""
            state:player_location(Room),
            state:inventory(Inv),
            findall(P, state:puzzle_solved(P), Puzzles),
            findall(key_dropped(Key, R), state:key_dropped(Key, R), DroppedKeys),
            findall(piece_dropped(Piece, Puzzle, R), state:piece_dropped(Piece, Puzzle, R), DroppedPieces)
        """))
        
        if not result:
            return None
            
        state = {
            'room': result[0]['Room'],
            'inventory': result[0]['Inv'],
            'solved_puzzles': result[0]['Puzzles'],
            'dropped_keys': result[0]['DroppedKeys'],
            'dropped_pieces': result[0]['DroppedPieces']
        }
        
        # Get all door states
        door_states = []
        door_query = list(self.prolog.query("facts:door(From, To, _), state:door_state(From, To, State)"))
        for door in door_query:
            door_states.append({
                'from': door['From'],
                'to': door['To'],
                'state': door['State']
            })
        state['doors'] = door_states
        
        return state

    def initialize_game(self, choice=1):
        """Initialize game state"""
        if choice == 1:
            print("Loading predefined game...")
            return bool(list(self.prolog.query("""
                facts:load_predefined_game,
                constraints:load_default_constraints,
                state:player_location(a),
                state:retractall(inventory(_)),
                state:assertz(inventory([]))
            """)))
        elif choice == 2:
            print("Creating custom game...")
            return bool(list(self.prolog.query("""
                facts:create_custom_game,
                constraints:create_custom_constraints,
                state:player_location(a),
                state:retractall(inventory(_)),
                state:assertz(inventory([]))
            """)))
        return False

    def check_win_condition(self):
        """Check if player has reached the final room"""
        result = list(self.prolog.query("facts:final_room(Room), state:player_location(Room)"))
        return bool(result)

    def get_available_actions(self):
        """Get available actions in current room"""
        actions = []
        current_room = self.player_location()
        
        # Check for keys in room
        keys = list(self.prolog.query(f"facts:key_in_room({current_room}, Key)"))
        for key in keys:
            actions.append(f"pick_key:{key['Key']}")
        
        # Check for dropped keys in room
        dropped_keys = list(self.prolog.query(f"state:key_dropped(Key, {current_room})"))
        for key in dropped_keys:
            actions.append(f"pick_key:{key['Key']}")
        
        # Check for puzzle pieces in room
        pieces = list(self.prolog.query(f"facts:piece_in_room({current_room}, Piece, Puzzle)"))
        for piece in pieces:
            actions.append(f"pick_piece:{piece['Piece']}")
        
        # Check for dropped pieces in room
        dropped_pieces = list(self.prolog.query(f"state:piece_dropped(Piece, Puzzle, {current_room})"))
        for piece in dropped_pieces:
            actions.append(f"pick_piece:{piece['Piece']}")
        
        # Check for movable objects
        objects = list(self.prolog.query(f"facts:object_in_room({current_room}, Object)"))
        for obj in objects:
            if not self.object_moved(obj['Object']):
                actions.append(f"move_object:{obj['Object']}")
        
        # Check for puzzles that can be solved
        puzzles = list(self.prolog.query(f"facts:puzzle_room(Puzzle, {current_room}), \+ state:puzzle_solved(Puzzle)"))
        for puzzle in puzzles:
            actions.append(f"solve_puzzle:{puzzle['Puzzle']}")
        
        # Check for doors that can be unlocked
        doors = list(self.prolog.query(f"facts:door({current_room}, OtherRoom, _), state:door_state({current_room}, OtherRoom, locked)"))
        for door in doors:
            actions.append(f"unlock_door:{door['OtherRoom']}")
        
        # Check for adjacent rooms
        adjacent = list(self.prolog.query(f"facts:door({current_room}, OtherRoom, _)"))
        for room in adjacent:
            door_state = self.door_state(current_room, room['OtherRoom'])
            if door_state == 'unlocked':
                actions.append(f"move_to:{room['OtherRoom']}")
        
        return actions
    
    def get_collected_items(self):
        """Return all collected keys and pieces"""
        print("entrego",self.collected_items)
        return {
            'keys': list(self.collected_items['keys']),
            'pieces': list(self.collected_items['pieces'])
        }

    def has_collected_key(self, key):
        """Check if a specific key has been collected"""
        return key in self.collected_items['keys']

    def has_collected_piece(self, piece):
        """Check if a specific puzzle piece has been collected"""
        return piece in self.collected_items['pieces']